### PR TITLE
[Python] Add new request types available in the OpenSky REST API.

### PR DESCRIPTION
**Pull request by marataj**
I recently found the OpenSky service, and it's great! I used it in my private data analysis projects. So far I have mainly worked with the REST API, but decided to adapt the python library to all request types that were needed. During the work, the code has been tested on Python 3.10 and 2.7 to keep compatibility, as assumed.
I'm happy to contribute to this project!

The scope of changes is the extension of the `OpenSkyApi` class with additional methods, that cover all types of requests available in the OpenSky REST API.

### CODE REVIEW

File: python/opensky_api.py [hunk 1]
```diff
@@ -336,4 +415,110 @@ def get_flighs_from_interval(self, begin: int, end: int) -> list[FlightData]:
         if states_json is not None:
             flights_list = [FlightData(list(entry.values())) for entry in states_json]
             return flights_list
-        return None
+        return []
+
+    def get_flighs_by_aircraft(self, icao24: str, begin: int, end: int) -> list[FlightData]:
```
Comment (jaluebbe) [inline python/opensky_api.py]: typo "flighs" -> "flights"

File: python/opensky_api.py [hunk 2]
```diff
@@ -336,4 +415,110 @@ def get_flighs_from_interval(self, begin: int, end: int) -> list[FlightData]:
...
+        )
+
+        if states_json is not None:
+            flights_list = [FlightData(list(entry.values())) for entry in states_json]
+            return flights_list
+        return []
+
+    def get_track_by_aircraft(self, icao24: str, t: int = 0) -> FlightTrack:
+        """
+        Retrieve flights for a certain airport which arrived within a given time interval [begin, end].
```
Comment (jaluebbe) [inline python/opensky_api.py]: The docstring does not fit. It is not about an airport.

File: python/opensky_api.py [hunk 3]
```diff
@@ -77,10 +106,15 @@ def __str__(self):
...
-      |  **time** - in seconds since epoch (Unix time stamp). Gives the validity period of all states. All vectors represent the state of a vehicle with the interval :math:`[time - 1, time]`.
-      |  **states** - a list of `StateVector` or is None if there have been no states received
+      |  **time**: `int` - in seconds since epoch (Unix time stamp). Gives the validity period of all states.
+        All vectors represent the state of a vehicle with the interval :math:`[time - 1, time]`.
+      |  **states**: `list` [`StateVector`] - a list of `StateVector` or is None if there have been no states received
     """
     def __init__(self, j):
+        """
+        Initializes the OpenSkyStates object
+        :param dict j: the json represents the state of the airspace as seen by OpenSky at a particular time
```
Comment (jaluebbe) [inline python/opensky_api.py:116]: Are you speaking of a JSON string or a Python dict? The character j might be confusing as it is being used for the imaginery unit in Python. I recommend a longer variable name.

File: python/opensky_api.py [hunk 4]
```diff
@@ -94,16 +128,140 @@ def __str__(self):
...
+    def __repr__(self):
+        return "FlightData(%s)" % repr(self.__dict__.values())
+
+    def __str__(self):
+        return pprint.pformat(self.__dict__, indent=4)
+
+
+class Waypoint(object):
+    """
+    Class that represents the singnle waypoint that is a basic part of flight trajectory:
```
Comment (jaluebbe) [inline python/opensky_api.py:191]: typo in "singnle"

File: python/opensky_api.py [hunk 5]
```diff
@@ -201,9 +387,134 @@ def get_my_states(self, time_secs=0, icao24=None, serials=None):
...
+        if states_json is not None:
+            return [FlightData(list(entry.values())) for entry in states_json]
+        return None
+
+    def get_flights_by_aircraft(self, icao24, begin, end):
+        """
+        Retrieves data of flights for certain aircraft and time interval.
+
+        :param str icao24: Unique ICAO 24-bit address of the transponder in hex string representation.
+            All letters need to be lower case
```
Comment (jaluebbe) [inline python/opensky_api.py:429]: missing "." at the end of he line

File: python/test_opensky_api.py [hunk 6]
```diff
@@ -101,6 +101,61 @@ def test_state_vector_parsing(self):
...
+        ]
+        values = [1675450754, 39.0912, -94.5794, 304, 91, False]
+        w = Waypoint(values)
+        for key, exp_val in zip(keys, values):
+            self.assertEqual(exp_val, getattr(w, key))
+
+    def test_flight_track_parsing(self):
+        entry = {'icao24': 'a01391',
+                 'callsign': 'N104AA  ',
+                 'startTime': 1675450754.0,
```
Comment (jaluebbe) [inline python/test_opensky_api.py:141]: is the time in the API response float or int?

File: python/opensky_api.py (hunk 3 above)
Comment (marataj) [reply python/opensky_api.py:116]: Dictionary is the right type.

File: python/opensky_api.py (hunk 4 above)
Comment (marataj) [reply python/opensky_api.py]: Done

File: python/opensky_api.py (hunk 5 above)
Comment (marataj) [reply python/opensky_api.py]: Done

File: python/opensky_api.py (hunk 1 above)
Comment (marataj) [reply python/opensky_api.py]: fixed with #[125f149](https://github.com/openskynetwork/opensky-api/pull/53/commits/125f14937864b1a601d4d7bd530063693339ba37)

File: python/opensky_api.py (hunk 2 above)
Comment (marataj) [reply python/opensky_api.py]: Done

File: python/test_opensky_api.py (hunk 6 above)
Comment (marataj) [reply python/test_opensky_api.py]: Good catch, `int` is the correct type here. 
Done. I fixed types also in waypoint representation

File: python/opensky_api.py [hunk 7]
```diff
@@ -137,24 +317,31 @@ def _check_rate_limit(self, time_diff_noauth, time_diff_auth, func):
...
 
     @staticmethod
     def _check_lon(lon):
         if lon < -180 or lon > 180:
-            raise ValueError("Invalid longitude {:f}! Must be in [-180, 180]".format(lon))
+            raise ValueError(
+                "Invalid longitude {:f}! Must be in [-180, 180].".format(lon)
+            )
 
-    def get_states(self, time_secs=0, icao24=None, serials=None, bbox=()):
```
Comment (jaluebbe) [inline python/opensky_api.py:147]: You removed the unused parameter "serials". As it is not mentioned in the official documentation this seems to be ok.

**Review by jaluebbe — COMMENTED** [review]
I did not test the modifications yet on my own. However, I could say at this point, that the existing functionality is not affected. The existing code as well as docstrings were reformatted. Some documentation was added.

File: python/opensky_api.py (hunk 7 above)
Comment (marataj) [reply python/opensky_api.py:147]: Yes, I removed this parameter in order to keep consistency with required parameters described in official docs.

**Comment by jaluebbe** [comment]
To clarify how to use your new class/methods, you may create the documentation for https://openskynetwork.github.io/opensky-api/python.html by editing https://github.com/openskynetwork/opensky-api/blob/master/docs/free/python.rst .

**Comment by marataj** [comment]
> To clarify how to use your new class/methods, you may create the documentation for https://openskynetwork.github.io/opensky-api/python.html by editing https://github.com/openskynetwork/opensky-api/blob/master/docs/free/python.rst .

Sure, I will complete the documentation for that.

**Comment by jaluebbe** [comment]
I looked into the documentation. It looks great. However, it is not clear where the user credentials are required and where not. Compare to the example for my states "api = OpenSkyApi(USERNAME, PASSWORD)".

**Comment by jaluebbe** [comment]
all tests passed on my machine.

**Pull request merged**